### DISCUSSION TITLE
Working on the backend with Nextjs ..  how to's...

### DISCUSSION THREAD

**Discussion by Elindo586** [vercel/next.js#37012] [Help]
I noticed on the Next.js template there is a folder with the name of API under the pages folder.

In a regular REACT app I normally put all the work in a separate server folder inside the root folder.. in that server folder is included all the dependencies and connections to database.
project here:
https://github.com/Elindo586/techunion/tree/main/server

Questions:
What is the reasoning for the Nextjs template to have an API folder under the Pages folder?  Can I just keep my server files separate on a server folder ?   I think it looks a bit cleaner that way..

After getting data to my backend, I need to go to the frontend and fetch that data to show it on my page....  and I think I need to change the code to getStaticProps (I think)... 

If I use getStaticProps.. would I see the data transfer from getStaticProps to the return() section during npm run dev?    How can I test it that it is passing data? 

The frontend page is this one before modifying it
https://github.com/Elindo586/techunion/blob/main/client/src/pages/Inventory.js

**Comment by icyJoseph** [comment]
The API routes are a back-end to your front-end. When the client needs to fetch data from an API, it can do so in two ways:

- directly from the client
- through an API route

In the first, the client needs to know the target URL, have credentials (API keys), and not be blocked by CORS.

In the second, the target URL and API keys exist only on the server, and are not seen by the client. CORS is less of an issue in this context too.

These API routes are not necessarily server routes, or these are actual endpoints that people can also reach, I can do go to `mydomain.com/api/hello` and get an answer from there.

These API routes have limitations in the size of data they can handle, and the time they can be alive, usually within 30 seconds. You are also not guaranteed to hit the same instance, even on successive calls, the server spins them up and down as it see necessary.

You can still have your server handling bunch elsewhere, and just implement the REST interface in API routes, that is, reject methods you don't want, get data from other servers, do some transformations if needed, respond with data, or errors that do not leak secrets, not implementation details, beyond the very necessary.

You can set cookies and do redirects from API routes. It's just great!


## GSSP and GSP

That being said, a few gotchas, GetServerSideProps and GetStaticProps, should not (and in fact GetStaticProps fails, while GetServerSideProps can work under severe and brittle hacks), reach for API routes, instead GetStaticProps and GSSP, should reach out to the 3rd party service directly.

A CRA/Vite app, on CSP, cannot access API routes, because it is just a stand-alone static HTML application with some JS and CSS.

A Next.js application has access to a server, and this enables a world of possibilities. IMHO, API routes are quite underrated! They are incredibly powerful!


## Answers

> What is the reasoning for the Nextjs template to have an API folder under the Pages folder? Can I just keep my server files separate on a server folder ? I think it looks a bit cleaner that way..

They are there to ease the load on the client. If your application needs to do some work with data, it could be done instead on an API route. API routes also have access to the Node API and can import pretty heavy modules without penalising the client initial loading times. 

They live under `pages/api`, because the file system routing makes it so that they are reachable at `/api/hello`, to give an example.

They can also be dynamic, and nested, and all of the other goodies from pages.

> After getting data to my backend, I need to go to the frontend and fetch that data to show it on my page.... and I think I need to change the code to getStaticProps (I think)...

In getStaticProps, you should get the data directly from the source, it is safe, nobody will see this code.  Check this tool, https://next-code-elimination.vercel.app/, in case you have doubt. And use it to make sure you are correct.

Whatever you return from getStaticProps is used as inputs for the page component. This is also a json payload that is included with the page, and used when navigating to the page on the client (when you are already on another page).

If the data risks becoming stale, you could also just fetch it from the client, by reaching to an API route. Though you'd miss out on SEO benefits. There's ISR and less preferably, GetServerSideProps, to handle that.

> If I use getStaticProps.. would I see the data transfer from getStaticProps to the return() section during npm run dev? How can I test it that it is passing data?

Yes, you can place logs if you want, or in `_app.tsx` log the `pageProps` property. At the end of the day, these are passed to the Page component. You'll just see it.

When you build the application a JSON file with the results o getStaticProps is created, and also bundled to the client.

**Comment by Elindo586** [comment]
OK... there is a lot of stuff for my brain to process at the moment.

Are you saying getStaticProps should connect directly to MongoDB and get the information from the collections?....  no need for cors mongoose express?...  where could I find a sample of that code? 


As of right now I am connecting my backend to MongoDB, getting the parts, and storing them in a different file.... then from my frontend I fetch the data...

I was/am thinking that getStaticProps could also fetch the data from my storing location?..

Currently to fetch data on the frontend I have this:
```
// fetch data from server
  const getInventory = async () => {
    await fetch("/api/parts")
      .then((response) => {
        if (!response.ok) {
          return response.statusText();
        }
        return response.json();
      })
      .then((invDataArr) => {
        partsList = invDataArr;
      })
      .catch((error) => console.log(error));
  };

```

**Reply by icyJoseph** [reply]
Hi,

There's not one-liner to get this right, I think. Reading the docs helps a lot though!

First, the [next.js examples page](https://nextjs.org/examples) contains a bunch of helpful starters, for instance, https://github.com/vercel/next.js/tree/canary/examples/with-mongodb-mongoose.

I've also given a talk about Next.js, which lives here: https://github.com/icyJoseph/next-js-react-gbg it contains slides and a demo-app, which implements most things you need. It even includes [an architecture ovefview](https://next-js-react-gbg.vercel.app/).

> You are **now**, actually doing the fun Next.js stuff 😃 

With all of that out of the way.

> Are you saying getStaticProps should connect directly to MongoDB and get the information from the collections?.... no need for cors mongoose express?... where could I find a sample of that code?

In a sense yes,  Next.js can handle creating the connection and all of that, on the server side.

However, getStaticProps, **runs only once at build time, and then it never runs again**. Talking **production** here, not dev mode.

There's a technique called **revalidation**, which allows you to re-run getStaticProps at runtime, see here https://nextjs.org/docs/basic-features/data-fetching/incremental-static-regeneration.

This will keep the data somewhat fresh. You can even trigger revalidation manually nowadays, https://nextjs.org/docs/basic-features/data-fetching/incremental-static-regeneration#on-demand-revalidation-beta.

You'd still need to **use a client or some API to make it easy to reach to the MongoDB**, but no additional server is needed.

You can also make an `api/route` that returns data from the database, but of course be careful with how you expose your database there. 

> As of right now I am connecting my backend to MongoDB, getting the parts, and storing them in a different file.... then from my frontend I fetch the data...

> I was/am thinking that getStaticProps could also fetch the data from my storing location?..

Not sure how this file works, but if it is on another server, and the file is exposed, for you to reach, then yeah, technically, it is possible for getStaticProps to also reach out for it. But that file must be available at build time! (when you run `yarn build` or when your CI runs it, either through the internet or file system)

## Data Fetching libraries

There's tools like SWR and React-Query, which help you to do an even more intriguing thing... You could, of course, do this with other libraries, or "vanilla", but why reinvent the wheel.

You can get the seed data from getStaticProps, and then use that as fallback data on the client. In turn SWR goes and fetches the latest data.

## SEO

If you render the page totally with data from getStaticProps, this is what bots, will see, until you deploy again (with different data). Do not assume that bots run JavaScript. So if the client updates the data once it renders, it is not guaranteed that a bot will see that.

This is where the revalidate option from above is very good, because bots will from time to time see new HTML.

GetServerSideProps is a bit of an extreme solution here, because it doesn't run at build time, and instead, only runs on every single request that reaches to the server. This will surely overload the amount of incoming connections to your database, so I'd avoid it in this case.

**Reply by Elindo586** [reply]
Oh... I didn't see this one.... lots of good info...   I think I will try to do the getStaticProps with revalidation...  the data wont change often, but whenever once a while does I want it to be changed somewhat automatically.

**Comment by Elindo586** [comment]
Maybe this ... I think it might save a lot of the backend code

https://www.mongodb.com/developer/how-to/nextjs-with-mongodb/
